packages/peregrine/lib/talons/RootComponents/Category/useCategory.js

```diff
@@ -0,0 +1,216 @@
+            return;
+        }
+
+        const filters = getFiltersFromSearch(search);
+
+        // Construct the filter arg object.
+        const newFilters = {};
+        filters.forEach((values, key) => {
+            newFilters[key] = getFilterInput(values, filterTypeMap.get(key));
+        });
+
+        // Use the category id for the current category page regardless of the
+        // applied filters. Follow-up in PWA-404.
+        newFilters['category_id'] = { eq: String(id) };
+
+        runQuery({
+            variables: {
+                currentPage: Number(currentPage),
+                id: Number(id),
+                filters: newFilters,
+                pageSize: Number(pageSize),
+                sort: { [currentSort.sortAttribute]: currentSort.sortDirection }
+            }
+        });
+    }, [
+        currentPage,
+        currentSort,
+        filterTypeMap,
+        id,
+        pageSize,
+        runQuery,
+        search
+    ]);
+
+    const totalPagesFromData = data
+        ? data.products.page_info.total_pages
+        : null;
+
+    useEffect(() => {
+        setTotalPages(totalPagesFromData);
+        return () => {
+            setTotalPages(null);
+        };
+    }, [setTotalPages, totalPagesFromData]);
+
+    // If we get an error after loading we should try to reset to page 1.
+    // If we continue to have errors after that, render an error message.
+    useEffect(() => {
+        if (error && !categoryLoading && currentPage !== 1) {
+            setCurrentPage(1);
+        }
+    }, [currentPage, error, categoryLoading, setCurrentPage]);
+
+    // Reset the current page back to one (1) when the search string, filters
+    // or sort criteria change.
+    useEffect(() => {
+        // We don't want to compare page value.
+        const prevSearch = new URLSearchParams(previousSearch.current);
+        const nextSearch = new URLSearchParams(search);
+        prevSearch.delete('page');
+        nextSearch.delete('page');
+
+        if (
+            prevSearch.toString() !== nextSearch.toString() ||
+            previousSort.current.sortAttribute.toString() !==
+                currentSort.sortAttribute.toString() ||
+            previousSort.current.sortDirection.toString() !==
+                currentSort.sortDirection.toString()
+        ) {
+            // The search term changed.
+            setCurrentPage(1);
+            // And update the ref.
+            previousSearch.current = search;
+            previousSort.current = currentSort;
+        }
+    }, [currentSort, previousSearch, search, setCurrentPage]);
+
+    const categoryData = categoryLoading && !data ? null : data;
+    const metaDescription =
+        data && data.category && data.category.meta_description
+            ? data.category.meta_description
+            : '';
+
+    const loading =
+        (introspectionCalled && !categoryCalled) ||
+        (categoryLoading && !data) ||
+        introspectionLoading;
```

Comment:
From the perspective of the UI component `loading` should be true if and only if we have do not have data and there are in flight requests.

You can simplify this:

```
const loading = !data && categoryLoading;
```

Reference [this example](https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/OrderHistoryPage/useOrderHistoryPage.js#L29-L49) to see how we use data and loading states to determine when to render the full page loading indicator versus the background loading indicator.

Reply:
It is done similar to [useSearchPage](https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/SearchPage/useSearchPage.js#L198), when only categoryLoading is involved, `noProductsFound` component flashes for a moment.

Reply:
Ok, please leave a comment indicating it at least, since it is an outlier.

Reply:
Done